### ISSUE TITLE
A trustful blockchain-based token economy to prevent bandwidth free-riding

### ISSUE THREAD

**Issue by devos50** [Tribler/tribler#3337]
This issue will be the main discussion place for our ongoing project: "A blockchain-based token economy to prevent bandwidth free-riding" (the title is open for discussion). The basic idea is to create a micro-economy within the Tribler platform for earning, spending and trading bandwidth tokens. This brings together various research topics, including blockchain-powered decentralized market, anonymous downloading and hidden seeding. Trustworthy behavior and participation should be rewarded while cheating should be punished. A basic policy should prevent users from selfishly consuming bandwidth without giving anything back. This directly addresses the tragedy-of-the-commons phenomena.

Our initial release should provide basic primitives to earn, trade and spend tokens. Our work could be extended with more sophisticated techniques like TrustChain record mixing, multiple identities, a robust reputation mechanism for tunnel selection, global consensus and verifiable public proofs (proof-of-bandwidth/proof-of-relay).

High-level overview:
![image](https://user-images.githubusercontent.com/325224/35796657-8674c372-0a5d-11e8-948e-3b45064c0492.png)

Initial architecture diagram:

![microeconomy](https://user-images.githubusercontent.com/1707075/34880991-81a79012-f7b2-11e7-891b-d73430583f52.png)

**Related issues:**
- (anonymous) payouts: #2571
- earn bandwidth tokens: #1842 
- spend bandwidth tokens: #3
- trade tokens: #2559
- small-scale mining pool: #3326 

**Related work:**
- joystream (http://joystream.co)
- privatix (https://privatix.io) - an ERC20 bandwidth token on top of the Ethereum blockchain. Their project has no code and a rather weak whitepaper.

**Open tasks (subject to change):**

Must have:
- [x] Token Mining. Earn tokens by donating bandwidth. Credit mining effortlessly converts idle bandwidth into tokens. The mining boosts the health of the network. Underseeded swarms and popular swarms are identified and helped out. Dead swarms and spam are correctly dealt with. Users need to select a channel with content to boost. This signal from the user prevents spam. In a later stage we will auto-boost swarms in general (implemented in #3390 by @egbertbouman).
- [x] Token mining by relaying Tor-like encrypted traffic (implemented in #3492). 
- [x] Get tokens for hidden seeding and being an exit node (implemented in #3492).
- [x] Market. Sell and buy on our integrated decentralised marketplace. Convert bandwidth tokens into Ethereum or Bitcoin. Your karma as a merchant grows with each successful trade. Part of our Internet-of-Trust research (#3502).
- [x] Coin wallet. Move coins between a Tribler PC and Android devices. Offline storage inside Android smartphones. Transfer securely and effortlessly using simple QR codes without the need for Internet access. Focus on the "you can see it, you can spend it" security model (implemented in #3358).
- [x] Basic verification experiment on DAS5. To verify the correctness and get initial result about the effectiveness of our bandwidth tokens, we should set up a basic DAS5 experiment that incorporates and tests all of the tasks above. (I've set up my own testnet with exit nodes, a tracker and Tribler runners on our Proxmox servers)
- [x] Basic freeride policy. When downloading, freeriders (users with a negative bandwidth token balance) should get lower priority until they have restored their balance again. See also #3573

Should have:
- [x] Write a reliable TrustChain crawler so we can get a dataset. It is important to get an overview of our network and activity, therefore we need a reliable crawler. Implemented in #3572: this crawler crawls from the genesis block onwards and tries to fill any gaps in trustchains when it can.
- [ ] Double-spend detection of TriblerChain records. A Client should be able to detect whether a malicious user has reused his prior records and keep a list of such events. Tribler should be able to act accordingly with users that (recently) performed a double-spend, like refusing services to them for a few days (https://github.com/Tribler/py-ipv8/pull/68).

**Comment by devos50** [comment]
High-level diagram:

![microeconomy](https://user-images.githubusercontent.com/1707075/34880991-81a79012-f7b2-11e7-891b-d73430583f52.png)

**Comment by devos50** [comment]
@synctext I've 'refactored' this issue as we discussed last week.

**Comment by devos50** [comment]
I've verified using an isolated exit node, an isolated Dispersy tracker and a single Tribler instance, that anonymous downloading is working well and the exit nodes are paid correctly. I will now scale up the experiment and add more nodes. This allows me to explore a basic freeride policy, in the case of oversupply and undersupply.

**Comment by devos50** [comment]
After #3502 is merged, the market is production-ready 👍

**Comment by synctext** [comment]
"Our live token economy: measuring and mitigating real-world attacks"
Detection of fraud on our live network and attacking our own network with live Sybil attacks could be The Big Next Paper... Cue "incremental personal PageRank". Merge trust of barter, voting, and trade transaction graphs.

**Comment by devos50** [comment]
Small game fallacy: http://unenumerated.blogspot.com/2015/05/small-game-fallacies.html?m=1

**Comment by devos50** [comment]
Current status:
- a TrustChain crawler is setup and running to fetch blocks created on the testnet. See https://github.com/devos50/py-ipv8/tree/trustchain_crawler for the code.
- an explorer has been setup to make it easier for me to debug what happens on the network. See http://130.161.119.212 (this only shows block on the testnet).
- worked on a testnet with some dedicated tunnel helpers.

**Comment by devos50** [comment]
Now that direct payouts are working (see #3776), I will focus on end-to-end integration experiments. I will start building these experiments in isolation but eventually, they will join the TrustChain testnet. Scientific goal: experimentally prove that free-riders are allocated less resources at exit nodes/relays.

**Comment by ichorid** [comment]
Oops, misclick on the Projects page... :face_with_head_bandage:

**Comment by devos50** [comment]
[Proof-of-Work as Anonymous Micropayment: Rewarding a Tor Relay](https://link.springer.com/chapter/10.1007%2F978-3-662-47854-7_27)

In this work, the authors propose a mechanism to reward Tor relay nodes, based on Proof-of-Work requirements by cryptocurrencies. Tor clients obtain Proof-of-Work shards and can turn them in later for preferential treatment.

They also propose a hierarchical token bucket algorithm to ensure QoS for "free" and "paying" clients. This is related to our slotted mechanism implemented in relay and exit nodes.

![Schermafbeelding 2019-08-13 om 14 56 19](https://user-images.githubusercontent.com/1707075/62943477-da082d80-bdda-11e9-834a-ba8641cc913a.png)

**Comment by synctext** [comment]
Thesis final chapter brainstorm:
- Trustful blockchain-based decentral markets
- Comparing numerous blockchain solutions (including EBSI)
- Trustworthy decentral exchange market for bandwidth tokens and cybercurrency
- Trustchain: measuring the emergence of trust
- Trustchain: measuring the emergence of trust in Markets, Money, and Media.
- BTC-ETH: Global decentral marketplace for cybercurrency

**Comment by devos50** [comment]
Final thesis chapter idea (as also briefly discussed yesterday): __Tribler: 15 years of research on managing the Internet commons__. Also, see [here](https://github.com/Tribler/paper-ideas/issues/1). This will be a journal paper.

![system_architecture](https://user-images.githubusercontent.com/1707075/87550951-f0380100-c6af-11ea-8add-8494191a6a16.png)

What problem did we solve? We design, implement and evaluate a fully decentralized infrastructure to address _free-riding behaviour in shared-resource systems_, oriented around three cardinal components: community accounting, reputation mechanisms and resource allocation. These three components together are capable of detecting fraudulent behaviour, e.g., record withholding and the Sybil Attack.

__Key contributions:__
- The system architecture of Tribler, our infrastructure for tragedy-of-the-commons experimentation.
- Large-scale experiments, exploring the effect of different reputation algorithms and scheduling policies.
- A longitudinal deployment trial, involving over 80.000 Internet users, demonstrating that Tribler effectively detects and punishes free-riding behaviour.

**Comment by synctext** [comment]
Related work on decentral marketplace with tokens: [FileBounty: Fair Data Exchange](https://arxiv.org/pdf/2008.11362.pdf)
![image](https://user-images.githubusercontent.com/325224/92234684-c2528a00-eeb2-11ea-8c61-2120fc127d6e.png)
Their system uses Semi-trusted third party, Merkle-Damgård and Sponge Construction, Boneh-Lynn-Shacham signature scheme, and obviously zkSNARKs. This is a fascinating theoretical result. However, "an Ethereum smart contract serves for dispute mediation" and all these complex components are not suitable for real-world usage at "Facebook" deployment levels.

**Comment by devos50** [comment]
I have started a write-up on this ecosystem [here](https://github.com/Tribler/tribler/wiki/The-design-of-a-trustworthy-ecosystem-around-media-sharing). Using this issue as a place-holder for the images in the document.

![architecture](https://user-images.githubusercontent.com/1707075/94149841-b1c17e00-fe78-11ea-9699-0cc31ed75de0.png)

**Comment by synctext** [comment]
Reacting to: https://github.com/Tribler/tribler/pull/5689 and https://github.com/Tribler/tribler/issues/5676. Much progress towards freeding prevention :1st_place_medal: 
- its a network crawler run by every Tribler peer
- the term "query" could imply a local SQL cmd
- "Adds functionality where the transactions of a random peer are periodically queried"
  - why collect random info?
  - we need to optimise collecting useful info
  - maximize "cache hit"
- This is a full network crawler, not just a remote query
- shows how the above wording can be confusing to even Tribler developers
- security analysis (this is just a path-finding version)
  - This is tentative version, starting point for incremental improvement
  - full pull-based 
- How many UDP packets into a single transaction
  - How many BAMI records collected?
  - How to fit into packets
  - How many seconds interval
  - How can we see database filling up?
  - is there a measurement plan?
    - ground truth (amount of records at each peer)
    - recall and precision (what are we seeing at single central crawler)
- Scientific grounding
  - surrounded by uncertainty, no ground truth
  - avoid single-event truthfulness
  - avoid the worst (avoid making a perfect ranking)
- Big Tech alternative storyline
  - mainstream computer science work?
  - _adversarial_ [streaming algorithm](https://en.wikipedia.org/wiki/Streaming_algorithm)
  - Need for results
  - Deployment numbers
  - What graphs?
  - mechanism for distributed power
  - "Too much power in the hands of the few"
- "Eventual Publishable Thesis Chapter" (EPTC)
- Exit nodes use reputation
  - observe and evaluate
  - data logging with respect for privacy

**Comment by devos50** [comment]
This issue has most likely been deferred in favour of tag-related improvements so I'm un-assigning myself for now 👍

**Comment by qstokkink** [comment]
It seems like this issue was part of completed PhD work. I'll close it.
